src/Symfony/Component/Translation/Formatter/IntlMessageFormatter.php

```diff
@@ -12,27 +12,50 @@
 namespace Symfony\Component\Translation\Formatter;
 
 use Symfony\Component\Translation\Exception\InvalidArgumentException;
+use Symfony\Component\Translation\Exception\LogicException;
 
 /**
  * @author Guilherme Blanco <guilhermeblanco@hotmail.com>
  * @author Abdellatif Ait boudad <a.aitboudad@gmail.com>
  */
-class IntlMessageFormatter implements MessageFormatterInterface
+class IntlMessageFormatter implements MessageFormatterInterface, IntlFormatterInterface
 {
+    private $hasMessageFormatter;
+    private $cache = array();
+
     /**
      * {@inheritdoc}
      */
     public function format($message, $locale, array $parameters = array())
     {
-        try {
-            $formatter = new \MessageFormatter($locale, $message);
-        } catch (\Throwable $e) {
-            throw new InvalidArgumentException(sprintf('Invalid message format (%s, error #%d).', intl_get_error_message(), intl_get_error_code()), 0, $e);
+        return $this->formatIntl($message, $locale, $parameters);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function formatIntl(string $message, string $locale, array $parameters = array()): string
+    {
+        if (!$formatter = $this->cache[$locale][$message] ?? null) {
+            if (!($this->hasMessageFormatter ?? $this->hasMessageFormatter = class_exists(\MessageFormatter::class))) {
+                throw new LogicException('Cannot parse message translation: please install the "intl" PHP extension or the "symfony/polyfill-intl-messageformatter" package.');
+            }
+            try {
+                $this->cache[$locale][$message] = $formatter = new \MessageFormatter($locale, $message);
```

Comment:
should we actually memoize all the formatters ? This increases memory usage for each message being formatted. In my experience, a page is much more likely to translate lots of different messages than lots of messages being the same.
What is the benefit in terms of speed when skipping the instantiation the second time (when using the intl extension. I don't care about the performance of the polyfill here, as the way to improve performance in such case would be to install ext-intl). Does it actually justify such memory leak ?

Reply:
I benched this:
`\MessageFormatter::formatMessage('Hello {name}', 'en', array('name' => 'Bob'));`
vs
`$formatter->format(array('name' => 'Bob'));`

the later is 3 times faster than the former and each $formatter object takes 226 bytes.
With that in mind, I think it's worth doing it: for run-once requests we don't care, and for long-running ones, the benefit is worth it IMHO.